### PR TITLE
Allocate sharded tensors

### PR DESCRIPTION
**Pull request by cowanmeg**
Allocates sharded tensor views with sharded axis of size 1.

Global sharded inputs are initialized with the sharded tensor. For example:
```
TensorView* tv = makeConcrete({2, 3, 4});
tv->mesh = DeviceMesh({0, 1});
tv->axis(0)->parallelize(ParallelType::DIDx);
```
will accept an input of size (1, 3, 4)

Intermediate tensors are also allocated the sharded sizes. For example:
```
TensorView tv1 = makeConcrete({2, 3,});
TensorView* tv2 = set(tv1);
tv2->axis(0)->parallelize(ParallelType::DIDx);
```
Will only allocate tv2 tensors with the sharded size {1, 3}. The previous PR https://github.com/NVIDIA/Fuser/pull/1450/files did not allocate the sharded axis, resulting in a sharded size of {3}.

TODO: We may switch back to not allocating the sharded axis. 
TODO: FusionExecutorCache is disabled in multidevice executor sharded size vs. unsharded size cause binding errors.

### CODE REVIEW

File: csrc/device_lower/pass/allocation.cpp [hunk 1]
```diff
@@ -163,13 +163,17 @@ class AllocationInserter : public kir::ExprMutator {
...
     for (const auto id : maybe_rfactor_domain) {
-      if (id->isReduction() || id->isStride() || id->isDeviceDim()) {
+      if (id->isReduction() || id->isStride()) {
         continue;
       } else if (id->isBroadcast()) {
         // No matter whether this broadcast is expanded or not, we always
         // allocate size 1
         alloc_dims.emplace_back(id->container()->oneVal());
         continue;
+      } else if (id->isDeviceDim()) {
```
Comment (samnordmann) [inline csrc/device_lower/pass/allocation.cpp]: Maybe add to the previous else if branch ? by replacing the current condition with `id->isBroadcast() || id->isDeviceDim()`

File: csrc/executor.cpp [hunk 2]
```diff
@@ -521,9 +521,10 @@ std::vector<int64_t> getContiguousStrides(
 
 // Infer the size and stride of each dimension
 std::pair<std::vector<int64_t>, std::vector<int64_t>> inferShape(
-  //interesting. I think it can be used as-is.
-// What to do with the strides ? Maybe we should assert (or Warn) that the strides are 0
-// From the implementation of how to get the strides, I think it's not a problem
+    // interesting. I think it can be used as-is.
+    // What to do with the strides ? Maybe we should assert (or Warn) that the
```
Comment (samnordmann) [inline csrc/executor.cpp]: oooh that was an old comment :) Let's maybe either remove it or not modidy it

File: csrc/executor.cpp [hunk 3]
```diff
@@ -843,11 +844,15 @@ class BackwardTraverseFromAllocToRFactor {
 // into a format whose dimensions are consistent with the rFactor domain of tv.
 // For example, if the rFactor domain is [I1, I2], and the allocation domain is
 // [I2*I1], then we will allocate as [I2*I1], then do a tensor.view(I2, I1).t()
-// to get a tensor whose semantics is [I1, I2] but its memory is [I2*I1]. interesting
-// Another example, if the rFactor domain is [I1*I2] and the allocation domain
-// is [I1, I2], then we will allocate as [I1, I2] and do a tensor.view(I1*I2) to
-// get a tensor whose semantics is [I1*I2] but memory is [I1,I2]
-at::Tensor transformOutputFromAllocationToRFactor( //maybe it is brutal to completely discard a Did axis. In case of a reduction?
+// to get a tensor whose semantics is [I1, I2] but its memory is [I2*I1].
+// interesting Another example, if the rFactor domain is [I1*I2] and the
```
Comment (samnordmann) [inline csrc/executor.cpp]: let's also undo this modif or remove the comment I added and forgot to remove (i.e. //maybe it is brutal to completely discard a Did axis. In case of a reduction?)

File: csrc/executor.cpp [hunk 4]
```diff
@@ -895,8 +900,11 @@ std::pair<std::vector<int64_t>, std::vector<int64_t>> inferShapeOfOutput(
   for (const auto id : tv->getMaybeAllocationDomain()) {
     if (id->isReduction() || id->isStride()) {
       continue;
+    } else if (id->isDeviceDim()) {
+      symbolic_sizes.push_back(id->container()->oneVal());
```
Comment (samnordmann) [inline csrc/executor.cpp:896]: this is OK to me but maybe it's worth mentionning it to Naoya during next meeting to check if its not in conflict with bcast dimensions (that are tagged as being of size 1 as well)

File: csrc/executor.cpp [hunk 5]
```diff
@@ -1537,7 +1545,8 @@ void FusionExecutor::initializeExecutorEntry(
   // All information is gathered. Save it to ExecutorEntry
   executor_entry.launch_params = launch_params;
   executor_entry.outputs = output_info;
-  executor_entry.intermediates = intermediates; // this where is initialized what we want
+  executor_entry.intermediates =
```
Comment (samnordmann) [inline csrc/executor.cpp]: let's undo this change?

File: csrc/executor.cpp [hunk 6]
```diff
@@ -1657,7 +1666,6 @@ std::vector<at::Tensor> FusionExecutor::runFusion(
     dumpFusionArgs(
         fusion_id_, args, launch_constraints, compile_params, outputs);
   }
-
```
Comment (samnordmann) [inline csrc/executor.cpp]: undo ?

File: csrc/multidevice/executor.cpp [hunk 7]
```diff
@@ -75,14 +77,15 @@ std::unordered_map<Val*, c10::IValue> MultiDeviceExecutor::allocateRecvBuffers(
   if (fusion_copy->outputs().empty()) {
     return {};
   }
-  unshard(fusion_copy.get());
-  FusionExecutorCache fec(std::move(fusion_copy));
-  auto buffers = fec.allocOutputSpace(global_inputs_IValues);
+  // TODO: Not working with FusionExecutorCache
```
Comment (samnordmann) [inline csrc/multidevice/executor.cpp:80]: ok, that perfectly makes sense. It makes sense to keep your change -- however there is the problem of the very slow compilation time when not using FusionExecutorCache. We really need to fix that

**Review by samnordmann — COMMENTED** [review]
Thanks Meghan ! :) 
That really looks great. Haven't finished reviewing, will finish tomorrow, I just jotted some minor notes for remainder in the meantime. But that's an awesome contribution.

If I am not mistaken this patch could also be merged to upstream -- maybe it would be good to prepare a PR for NVIDIA repo as soon as we merge this one here.

File: test/test_multidevice_ir.cpp [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+#include <fusion_segmenter.h>
+#include <gtest/gtest.h>
+#include <multidevice/executor.h>
+#include <multidevice/utils.h>
+#include <ops/all_ops.h>
+#include <test/multidevice.h>
+#include <test/validator.h>
+
+namespace nvfuser {
+TEST_F(MultiDeviceTest, ShardOuterAxisConcrete) {
```
Comment (samnordmann) [inline test/test_multidevice_ir.cpp]: Can you explain what we want to test here ?

File: test/test_multidevice_ir.cpp [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+  auto x = at::randn({num_devices, 3}, tensor_options);
+  std::vector<c10::IValue> inputs = {x};
+  auto ref_outputs = at::sum(x * 4, {0});
+
+  MultiDeviceExecutor runtime(std::move(fusion), *communicator);
+  auto outputs = runtime.runWithInput(inputs);
+  testValidate(
+      runtime.fusion(), outputs, inputs, {ref_outputs}, __LINE__, __FILE__);
+}
+TEST_F(MultiDeviceTest, ShardOuterAxis) {
```
Comment (samnordmann) [inline test/test_multidevice_ir.cpp]: What is the difference with the previous test?
Can it be made a parametrized test, to avoid duplication?
If the only difference is how we define tv0, then a bool parameter can be used and smth like:
`TensorView* tv0 = bool_param ? makeContigTensor(2) : makeConcreteTensor({num_devices, 3});

File: test/test_multidevice_pipeline.cpp [hunk 10]
```diff
@@ -184,10 +186,13 @@ TEST_P(PipelineTestTwoStages, Communication) {
   if (is_stage0_sharded) {
     tv0->axis(0)->parallelize(ParallelType::DIDx);
     tv1->axis(0)->parallelize(ParallelType::DIDx);
+    // Shard outermost-axis of input
+    input_sizes[0] = 1;
```
Comment (samnordmann) [inline test/test_multidevice_pipeline.cpp]: alternatively, we could also remove this line and define `first_axis_extent` as follows:
```
int first_axis_extent = is_stage0_sharded? 1: 3;
```
it would reduce the number of lines.
Wdyt?

Maybe we could also get rid of the block defining`second_axis_extent` ?

File: test/test_multidevice_ir.cpp (hunk 8 above)
Comment (samnordmann) [reply test/test_multidevice_ir.cpp]: Maybe we could add a check that the sharded dimension in the input is indeed of size 1 ?

File: test/test_multidevice_ir.cpp [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+  std::vector<c10::IValue> inputs = {x};
+  auto ref_outputs = at::sum(x * 4, {0});
+
+  MultiDeviceExecutor runtime(std::move(fusion), *communicator);
+  auto outputs = runtime.runWithInput(inputs);
+  testValidate(
+      runtime.fusion(), outputs, inputs, {ref_outputs}, __LINE__, __FILE__);
+}
+
+inline at::Tensor shardInputTensor(at::Tensor tensor, int deviceId) {
```
Comment (samnordmann) [inline test/test_multidevice_ir.cpp]: I think there is a small issue here. the index for slicing shouldn't be deviceId but the index of deviceId in the device mesh associated with the input. Here it works because, by chance, with the chosen device mesh, those two indices are equal.

But consider the device mesh
dm = {2,1}
then, device 1 should get the slice 0
device 2 should get the slice 1
and device 0 and others should get ... ? probably an empty vector. But from what I understand the system will complain if a has an empty vector.

Maybe this should be fixed, to make sure that tensors on devices where it is not sharded are empty tensors
Or just we need to make sure (with an assert or a comment) that this utility function is only used when the device mesh is the full range of ranks. 

What do you think ?

File: csrc/device_lower/pass/allocation.cpp (hunk 1 above)
Comment (samnordmann) [reply csrc/device_lower/pass/allocation.cpp]: Also, I don't remember what was our conclusion about allocating with size 1 vs. not allocating..?
I see that allocating with size 1 implies to squeeze/unsqueeze the tensors, however I can't remember what is the benefit

File: test/test_multidevice_ir.cpp (hunk 11 above)
Comment (samnordmann) [reply test/test_multidevice_ir.cpp]: thats a detail but in term of nvFuser's codestyle, every static declaration (i.e. that is not linked with an external file) should be placed in an anonymous namespace

**Review by samnordmann — COMMENTED** [review]
If a tensor is not sharded on a given device, maybe it shouldn't be allocated at all right ?
If, say, a given input is not sharded on device 0, then don't we want this device to concretize the input as an empty tensor ?

See https://github.com/samnordmann/Fuser/pull/6/files#r1447535136

File: csrc/expr_evaluator.cpp [hunk 12]
```diff
@@ -141,6 +141,13 @@ void ExpressionEvaluator::bind_(
             i);
         bind_(
             rfactor_domain[i]->expandedExtent(), t.size(i), evaluate_validate);
+      } else if (
+          rfactor_domain[i]->isDeviceDim() &&
+          rfactor_domain[i]->getMaybeExpandedExtent()->isConst()) {
```
Comment (samnordmann) [inline csrc/expr_evaluator.cpp]: nit: I think `getMaybeExpandedExtent()` can be replaced by `extent()` here (because we are in the else branch)

File: csrc/device_lower/pass/allocation.cpp (hunk 1 above)
Comment (cowanmeg) [reply csrc/device_lower/pass/allocation.cpp]: We might need the device dimension allocated to handle splits.
```
tv0 = makeContig(2); // (n, m)
tv0->axis(0)->split(ndevices, false); // (ndevices, n//ndevices, m)
tv0->axis(0)->parallelize(DIDx);
```
will allocate a 2D tensor with size (n//ndevices, m)

File: test/test_multidevice_pipeline.cpp (hunk 10 above)
Comment (cowanmeg) [reply test/test_multidevice_pipeline.cpp]: For concrete tensors, the test passes the unsharded size into makeConcreteTensorView. We might want this to be the sharded size though.

File: test/test_multidevice_ir.cpp (hunk 8 above)
Comment (cowanmeg) [reply test/test_multidevice_ir.cpp]: This is just a simple test for checking that sharding the outermost axis is working for both concrete and contig tensorviews. It's a bit of a duplicate with the PipelineTwoStageTests, but it's a bit easier to debug.

File: csrc/device_lower/pass/allocation.cpp (hunk 1 above)
Comment (samnordmann) [reply csrc/device_lower/pass/allocation.cpp]: Why would it help then to have (1, n//ndevices, m) instead of (n//ndevices, m) ? I am not sure to understand why it is necessary

File: test/test_multidevice_pipeline.cpp (hunk 10 above)
Comment (samnordmann) [reply test/test_multidevice_pipeline.cpp]: Ok I see ! Well spotted :) So it means that with this patch we break the support for symbolic shapes and only support concrete tensors

File: test/multidevice.h [hunk 13]
```diff
@@ -59,6 +59,11 @@ class MultiDeviceTest : public NVFuserTest {
   std::vector<std::pair<const std::string, std::chrono::time_point<std::chrono::high_resolution_clock>>> times;
 };
 
+class ShardingTest
+    : public MultiDeviceTest,
```
Comment (samnordmann) [inline test/multidevice.h]: It's up to you but you can also declare the class in the `test_multidevice_sharding.cpp` file

**Comment by samnordmann** [comment]
If we decide to only support concrete sizes (which I think we should do), it means that we can fully remove the garbage allocation by allocating empty tensors on device where the tensors is not sharded.

Example:

```
Fusion fusion;
FusionGuard fg(&fusion);

auto tv = makeConcrete({2, 3, 4});
tv->mesh = DeviceMesh({0, 1});
tv->setDeviceMesh({0}
fusion.addInput(tv)
...
```
then, the concrete input associated to tv on devices 1,2,3... should be an empty vector.
Since the size is concrete, the allocator will be able to figure out the extent of all the tensors, so it there will be no problem.

If we want to proceed this way, it means that the way we bind axis to their concrete extent must take the device mesh into account, and we should accept empty tensors in devices where the tensor is not sharded on

File: csrc/device_lower/pass/allocation.cpp (hunk 1 above)
Comment (cowanmeg) [reply csrc/device_lower/pass/allocation.cpp]: I was thinking we allocate the same number of dimensions as the tensorview. So the above example would allocate (n//ndevices, m). 

```
tv0 = makeContig(2); // (n, m)
tv0->axis(0)->parallelize(DIDx);
```
Then, this case, which assumes n == ndevices, would allocate (1, m) instead of (m).

**Comment by cowanmeg** [comment]
> If we decide to only support concrete sizes (which I think we should do), it means that we can fully remove the garbage allocation by allocating empty tensors on device where the tensors is not sharded.
> 
> Example:
> 
> ```
> Fusion fusion;
> FusionGuard fg(&fusion);
> 
> auto tv = makeConcrete({2, 3, 4});
> tv->mesh = DeviceMesh({0, 1});
> tv->setDeviceMesh({0}
> fusion.addInput(tv)
> ...
> ```
> 
> then, the concrete input associated to tv on devices 1,2,3... should be an empty vector. Since the size is concrete, the allocator will be able to figure out the extent of all the tensors, so it there will be no problem.
> 
> If we want to proceed this way, it means that the way we bind axis to their concrete extent must take the device mesh into account, and we should accept empty tensors in devices where the tensor is not sharded on

That's a good idea. We can get a lot more features in with concrete shapes then extend it later for dynamic shape.

File: test/multidevice.h (hunk 13 above)
Comment (cowanmeg) [reply test/multidevice.h]: good idea, it will be more self contained.

File: test/test_multidevice_sharding.cpp [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+  MultiDeviceExecutor runtime(std::move(fusion), *communicator);
+  auto outputs = runtime.runWithInput(inputs);
+  testValidate(
+      runtime.fusion(), outputs, inputs, {ref_outputs}, __LINE__, __FILE__);
+}
+
+INSTANTIATE_TEST_SUITE_P(
+    OutermostAxis,
+    ShardingTest,
+    ::testing::Values(true, false)
```
Comment (samnordmann) [inline test/test_multidevice_sharding.cpp:116]: nit: `::testing::Bool()`

**Review by samnordmann — APPROVED** [review]
